bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -710,35 +727,70 @@ private Runnable getStorer(final byte[] data, final int readingOffset, final int
         @Override
         public void run()
         {
-          int i = storage.store(identifier, data, readingOffset, writingOffset);
-          if (i == 0) {
+          if (uniqueIdentifier == 0) {
+            uniqueIdentifier = storage.store(identifier, data, readingOffset, writingOffset);
+          }
+          if (uniqueIdentifier == 0) {
             logger.warn("Storage returned unexpectedly, please check the status of the spool directory!");
           }
           else {
             synchronized (Block.this) {
-              Block.this.uniqueIdentifier = i;
-              if (refCount == 0) {
+              if (refCount.get() == 0) {
+                //logger.debug("release block {} to disk", Block.this);
+                freeBuffers.offer(Block.this.data);
                 Block.this.data = null;
+                numberOfInMemBlockPermits++;
+              } else {
+                //logger.debug("Keeping Data List Memory Block {} due to {} references.", Block.this, refCount);
               }
             }
+            resumeSuspendedClients();
           }
         }
-
       };
     }
 
-    synchronized void release(boolean wait)
+    protected void release(boolean wait)
     {
-      if (--refCount == 0 && storage != null) {
-        if (uniqueIdentifier != 0) {
-          data = null;
-          return;
+      if (refCount.decrementAndGet() == 0 && storage != null) {
+        assert (Block.this.next != null);
+        final Runnable storer = getStorer(data, readingOffset, writingOffset, storage);
+        if (wait && numberOfInMemBlockPermits == 0) {
+          //logger.info("Releasing Data List Memory Block {}", this);
+          storer.run();
+        } else if (numberOfInMemBlockPermits < MAX_COUNT_OF_INMEM_BLOCKS/2) {
```

Comment:
Can we make this configurable? I think we should allow the factor of 2 to be configured and default to 2 so that we can allow more flexibility in background process running to spool blocks.